---
title: "Variants Analysis"
subtitle: "211001_APOE_Mm dataset"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Figure 4 - DAR along chromosome 7

```{r}
png(
  filename = "~/phd/publications/fisher_pub/figures/main/dar_mouse.png",
  width = 32,
  height = 24,
  units = "cm",
  res = 300
)
```

```{r}
ind <- 2
chr <- 7
genes_plot <- genes[deGenes[[ind]]$gene_id]
mutation <- genes[genes$gene_name == "Apoe"]
dist <- allele_dist[[ind]]
dist_type <- "b"
dist_trans <- 1
highlight_genes <- TRUE
lfc_plot <- lfc[[ind]]
lfc_type <- "heatmap"
lfc_trans <- 1/3

tracks <- list()

if (length(mutation)) {
  mutation <- mutation %>%
    plyranges::filter(seqnames == chr) %>%
    plyranges::mutate(symbol = gene_name)
  mutant_track <- GeneRegionTrack(
    range = mutation,
    transcriptAnnotation = "symbol",
    col = "white",
    fill = "white",
    showTranscriptId = TRUE,
    fontcolor.group = "red",
    cex.group = 0.6,
    size = 0.5,
    name = NULL
  )
  tracks <- append(tracks, mutant_track)
}

axis_track <- GenomeAxisTrack(
  add53 = TRUE,
  add35 = TRUE,
  name = paste0("Chr", chr),
  showTitle = TRUE,
  size = 1
)
if (length(mutation)) {
  axis_track <- HighlightTrack(
    trackList = list(axis_track),
    range = mutation
  )
}
tracks <- append(tracks, axis_track)

if (length(genes_plot)) {
  genes_plot <- genes_plot %>%
    plyranges::filter(seqnames == chr) %>%
    plyranges::mutate(symbol = gene_name)
  gene_track <- AnnotationTrack(
    range = genes_plot,
    name = "DE",
    shape = "box",
    fill = "darkgray",
    group = genes_plot$gene_name,
    groupAnnotation = "group",
    fontcolor.group = 1,
    cex.group = 0.8,
    size = 0.4
  )
  tracks <- append(tracks, gene_track)
}

dist_track <- dist[,"diversity"] %>%
  plyranges::filter(seqnames == chr) %>%
  DataTrack(
    type = dist_type,
    name = "DAR",
    size = 8,
    window = -1,
    windowSize = 1,
    cex = 0.4,
    col = "grey20",
    col.axis = "black",
    yTicksAt = seq(0, 1, 0.1),
    ylim = c(0, 1),
    transformation = function(x){x^(dist_trans)}
  )
if (highlight_genes) {
  if (all(length(genes_plot) & length(mutation))) {
    ranges <- c(genes_plot, mutation)
  } else if (length(genes_plot)) {
    ranges <- genes_plot
  } else if (length(mutation)) {
    ranges <- mutation
  }
  dist_track <- HighlightTrack(
    trackList = list(dist_track),
    range = c(genes_plot, mutation),
    col = c(rep("#ffe0e0", length(genes_plot)), rep("red", length(mutation))),
    fill = c(rep("#ffe0e0", length(genes_plot)), rep("red", length(mutation)))
  )
}
tracks <- append(tracks, dist_track)

# if (length(lfc_plot)) {
#   lfc_track <- lfc_plot[,"lfc_rollmean"] %>%
#     plyranges::filter(seqnames == chr) %>%
#     DataTrack(
#       type = lfc_type,
#       name = paste0("abs.logFC^(1/3)"),
#       size = 2,
#       window = -1,
#       windowSize = 1,
#       col.axis = "black",
#       showColorBar = TRUE,
#       gradient = viridis(100, option = "B"),
#       transformation = function(x){x^(lfc_trans)}
#     )
#   tracks <- append(tracks, lfc_track)
# }

plotTracks(
  trackList = tracks,
  cex.main = 1,
  cex.title = 0.6,
  col.title = "black",
  background.title = "white"
)
```

```{r}
dev.off()
```

# Supp 2 - Elastic window sizes

```{r}
lapply(seq_along(allele_dist), function(ind){
  allele_dist[[ind]] %>%
    as_tibble() %>%
    dplyr::filter(!is.na(win_size)) %>%
    mutate(
      comparison = factor(names(allele_dist[ind]), levels = names(allele_dist))
    )
}) %>%
  bind_rows() %>%
  mutate(comparison = str_replace(comparison, "_", " ")) %>%
  ggplot(aes(win_size)) +
  geom_histogram(
    bins = 100,
    colour = "black",
    fill = "grey70"
  ) +
  scale_x_log10(labels = function(x){x/1000}) +
  facet_wrap(~comparison, ncol = 2) +
  labs(
    x = "Window size (kb)",
    y = "Count"
  )
```

```{r}
ggsave(
  "~/phd/publications/fisher_pub/figures/supp/window_sizes.png",
  device = "png",
  width = 18,
  height = 12,
  units = "cm"
)
```

# Supp 3 - Cumulative distribution of DAR

```{r}
allele_dist[[2]] %>%
  split(seqnames(.)) %>%
  lapply(function(chr){
    tibble(
      chromosome = as.vector(seqnames(chr)),
      div = chr$diversity
    )
  }) %>%
  purrr::reduce(rbind) %>%
  mutate(
    chromosome = factor(
      chromosome,
      # Draw mutant chr on top
      levels = unique(c(
        chromosome[chromosome != "7"],
        chromosome[chromosome == "7"]
      ))
    ),
    mutant_chromosome = chromosome == "7"
  ) %>%
  ggplot(aes(x = div, group = chromosome, colour = mutant_chromosome)) +
  stat_ecdf() +
  scale_x_continuous(breaks = seq(0, 1, 0.1)) +
  scale_y_continuous(breaks = seq(0, 1, 0.1)) +
  scale_colour_manual(values = c("grey30", "red")) +
  labs(
    x = "DAR",
    y = "F(DAR)"
  ) +
  guides(
    colour = guide_legend(title = "Mutation-bearing\nchromosome", reverse = TRUE)
  )
```

```{r}
ggsave(
  "~/phd/publications/fisher_pub/figures/supp/darCumDist_apoe2v3f.png",
  device = "png",
  width = 18,
  height = 10,
  units = "cm"
)
```