<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>Differential Gene Expression Analysis</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">211001_APOE_Mm</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="de.html">DE</a>
</li>
<li>
  <a href="dar.html">DAR</a>
</li>
<li>
  <a href="enrichment.html">Enrichment</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/baerlachlan/211001_APOE_Mm">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Differential Gene Expression Analysis</h1>
<h3 class="subtitle">211001_APOE_Mm dataset</h3>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">22 March, 2023</h4>

</div>


<div id="setup" class="section level1">
<h1>Setup</h1>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(edgeR)
  library(fgsea)
  library(pheatmap)
  library(cqn)
  library(DT)
  library(htmltools)
  library(gtools)
})</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores &lt;- availableCores() - 1</code></pre>
<pre class="r"><code>source(&quot;~/bioinformatics/bioToolkit/lbFuncs.R&quot;)</code></pre>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<pre class="r"><code>ens_species &lt;- &quot;Mus musculus&quot;
ens_release &lt;- &quot;104&quot;
ens_assembly &lt;- &quot;GRCm39&quot;</code></pre>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == ens_species) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ahId &lt;- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb &lt;- ah[[ahId]]</code></pre>
<pre class="r"><code>chrInfo &lt;- getChromInfoFromEnsembl(ens_assembly, release = ens_release) %&gt;%
  dplyr::filter(coord_system == &quot;chromosome&quot;)
## Sort by numbers then letters
primary_chrs &lt;- chrInfo$name %&gt;%
  mixedsort()
## Push MT chr to end
primary_chrs &lt;- c(
  primary_chrs[primary_chrs != &quot;MT&quot;],
  primary_chrs[primary_chrs == &quot;MT&quot;]
)</code></pre>
<pre class="r"><code>transcripts &lt;- transcripts(ensDb, filter = SeqNameFilter(primary_chrs))
txLen &lt;- exonsBy(ensDb, &quot;tx&quot;, filter = SeqNameFilter(primary_chrs)) %&gt;%
  width() %&gt;%
  lapply(sum) %&gt;%
  unlist(use.names = TRUE) %&gt;%
  enframe(name = &quot;tx_id&quot;, value = &quot;length&quot;)
mcols(transcripts) &lt;- mcols(transcripts)[
  c(&quot;tx_id&quot;, &quot;gene_id&quot;, &quot;gc_content&quot;)
] %&gt;%
  as.data.frame() %&gt;%
  left_join(txLen)</code></pre>
<pre class="r"><code>geneGc &lt;- transcripts %&gt;%
  mcols() %&gt;%
  as_tibble() %&gt;%
  group_by(gene_id) %&gt;%
  summarise(
    gc_content = sum(gc_content*length) / sum(length),
    length = ceiling(median(length))
  )</code></pre>
<pre class="r"><code>genes &lt;- genes(ensDb, filter = SeqNameFilter(primary_chrs))
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
] %&gt;%
  as.data.frame() %&gt;%
  left_join(geneGc)</code></pre>
<p>An <code>EnsDb</code> object was obtained for Ensembl release 104
with the <code>AnnotationHub</code> package. This contained the
information required to set up gene and transcript annotations.
Gene-level estimates of GC content and length were also generated which
are required for assessing potential biases in downstream analysis.</p>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<pre class="r"><code>metadata &lt;- read_tsv(here(&quot;misc/SYNAPSE_METADATA_MANIFEST.tsv&quot;)) %&gt;%
  left_join(read_csv(here(&quot;misc/metaAPOE.csv&quot;))) %&gt;%
  dplyr::select(
    sample = specimenID, species, genotypeBackground, litter, dateBirth,
    dateDeath, genotype = Genotype, sex = Sex, age = Age, lane, basename = name,
    modelSystemName, individualID, study
  ) %&gt;%
  dplyr::filter(str_detect(sample, &quot;_3M_&quot;)) %&gt;%
  mutate(basename = str_remove(basename, &quot;.bam_R(1|2).fastq.gz&quot;)) %&gt;%
  distinct(sample, .keep_all = TRUE) %&gt;%
  mutate(
    group = as.factor(paste0(genotype, &quot;_&quot;, age, &quot;_&quot;, sex)),
    genotype = as.factor(genotype)
  ) %&gt;%
  dplyr::arrange(genotype, group)</code></pre>
<pre class="r"><code>genoCols &lt;- metadata$genotype %&gt;%
  unique() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set1&quot;) %&gt;%
  setNames(unique(metadata$genotype))</code></pre>
</div>
<div id="counts" class="section level2">
<h2>Counts</h2>
<pre class="r"><code>counts &lt;- read_tsv(here(&quot;analysis-de/results/03_counts/counts.out&quot;), comment = &quot;#&quot;) %&gt;%
  dplyr::select(-c(Chr, Start, End, Strand, Length)) %&gt;%
  set_colnames(basename(colnames(.))) %&gt;%
  set_colnames(str_remove(colnames(.), &quot;\\.FCHJKY5BBXX.+&quot;))</code></pre>
<pre class="r"><code>minLib &lt;- counts %&gt;%
  dplyr::select(-Geneid) %&gt;%
  colSums() %&gt;%
  min()
# minCPM &lt;- as.vector(cpm(10, lib.size = minLib))
minCPM &lt;- 1
minSamples &lt;- metadata %&gt;%
  group_by(genotype) %&gt;%
  summarise(n = n()) %&gt;%
  pull(n) %&gt;%
  min()
dgeList &lt;- counts %&gt;%
  as.data.frame() %&gt;%
  column_to_rownames(&quot;Geneid&quot;) %&gt;%
  .[rowSums(cpm(.) &gt;= minCPM) &gt;= minSamples,] %&gt;%
  DGEList(
    samples = tibble(sample = colnames(.)) %&gt;%
      left_join(metadata),
    genes = genes[rownames(.),] %&gt;%
      as_tibble() %&gt;%
      dplyr::select(chromosome = seqnames, everything()) %&gt;%
      mutate(chromosome = factor(chromosome, levels = primary_chrs))
  ) %&gt;%
  calcNormFactors()</code></pre>
<pre class="r"><code>dgeList_path &lt;- here(&quot;files/dgeList.Rds&quot;)
if (!file.exists(dgeList_path)) {
  saveRDS(dgeList, dgeList_path)
}</code></pre>
<p>Gene-level count data was loaded from <code>featureCounts</code> and
imported into <code>R</code> as a <code>DGEList</code> object. Genes
were removed from the counts matrix if they had less than 1 CPM (~10
reads) in more than 32 samples. This meant that each remaining gene must
have assigned reads detected in at least one genotype group.</p>
<p>14755 genes remained for further analysis with library sizes ranging
from 57,605,573 to 109,822,995.</p>
<pre class="r"><code>cpm(dgeList, log = TRUE) %&gt;%
  as.data.frame() %&gt;%
  pivot_longer(
    cols = everything(),
    names_to = &quot;sample&quot;,
    values_to = &quot;logCPM&quot;
  ) %&gt;%
  left_join(metadata) %&gt;%
  ggplot(aes(logCPM, colour = genotype, group = sample)) +
  geom_density() +
  scale_colour_manual(values = genoCols) +
  labs(
    x = &quot;logCPM&quot;,
    y = &quot;Density&quot;,
    colour = &quot;Genotype&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-1-1.png" alt="*Expression density plots for all samples after filtering*" width="768" />
<p class="caption">
<em>Expression density plots for all samples after filtering</em>
</p>
</div>
</div>
</div>
<div id="initial-de-analysis" class="section level1">
<h1>Initial DE analysis</h1>
<div id="pca" class="section level2">
<h2>PCA</h2>
<pre class="r"><code>cpm &lt;- dgeList %&gt;%
  cpm(log = TRUE)</code></pre>
<pre class="r"><code>pca &lt;- cpm %&gt;%
  t() %&gt;%
  prcomp()
pcaVars &lt;- percent_format(0.1)(summary(pca)$importance[&quot;Proportion of Variance&quot;,])</code></pre>
<pre class="r"><code>pca$x %&gt;%
  as.data.frame() %&gt;%
  rownames_to_column(&quot;sample&quot;) %&gt;%
  left_join(metadata) %&gt;%
  as_tibble() %&gt;%
  ggplot(aes(PC1, PC2, colour = genotype, fill = genotype, label = sample)) +
  geom_point(size = 2) +
  geom_text_repel(show.legend = FALSE) +
  stat_ellipse(geom = &quot;polygon&quot;, alpha = 0.05, show.legend = FALSE) +
  guides(fill = &quot;none&quot;) +
  scale_colour_manual(values = genoCols) +
  labs(
    x = paste0(&quot;PC1 (&quot;, pcaVars[[&quot;PC1&quot;]], &quot;)&quot;),
    y = paste0(&quot;PC2 (&quot;, pcaVars[[&quot;PC2&quot;]], &quot;)&quot;),
    colour = &quot;Genotype&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-4-1.png" alt="*Principal Component Analysis of gene-level counts*" width="768" />
<p class="caption">
<em>Principal Component Analysis of gene-level counts</em>
</p>
</div>
</div>
<div id="design" class="section level2">
<h2>Design</h2>
<pre class="r"><code>design &lt;- model.matrix(~0 + group, data = dgeList$samples) %&gt;%
  set_colnames(str_remove(colnames(.), &quot;group&quot;))</code></pre>
<pre class="r"><code>contrasts &lt;- makeContrasts(
  APOE2v3_female = APOE2_3M_female - APOE3_3M_female,
  APOE2v3_male = APOE2_3M_male - APOE3_3M_male,
  APOE4v3_female = APOE4_3M_female - APOE3_3M_female,
  APOE4v3_male = APOE4_3M_male - APOE3_3M_male,
  levels = colnames(design)
)</code></pre>
<pre class="r"><code>pheatmap(
  design[metadata$sample,],
  cluster_cols = FALSE,
  cluster_rows = FALSE,
  color = c(&quot;white&quot;, &quot;grey50&quot;),
  annotation_row = dgeList$samples[&quot;genotype&quot;],
  annotation_colors = list(genotype = genoCols),
  angle_col = &quot;315&quot;,
  legend = FALSE
)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-6-1.png" alt="*Visualisation of the design matrix*" width="576" />
<p class="caption">
<em>Visualisation of the design matrix</em>
</p>
</div>
</div>
<div id="model-fitting" class="section level2">
<h2>Model fitting</h2>
<p>Genewise negative binomial generalized linear models were using the
<code>glmFit()</code> function of the <code>edgeR</code> package.
Likelihood ratio tests were then performed for the coefficients in the
model with <code>glmLRT()</code>. Here the null hypothesis, <span
class="math inline">\(H_0\)</span>, is testing that the true values of
the coefficients are equal to 0. Due to the subtle effects of the
mutation outlined in the PCA, genes were considered to be DE if they had
an FDR-adjusted p-value &lt; 0.05, as opposed to more stringent error
rate controls such as the Bonferroni method.</p>
<pre class="r"><code>alpha &lt;- 0.05
minLfc &lt;- 1
fit &lt;- estimateDisp(dgeList, design) %&gt;%
  glmFit()
topTables &lt;- colnames(contrasts) %&gt;%
  sapply(function(x){
    glmLRT(fit, contrast = contrasts[,x]) %&gt;%
      topTags(n = Inf) %&gt;%
      .[[&quot;table&quot;]] %&gt;%
      as_tibble() %&gt;%
      arrange(PValue) %&gt;%
      dplyr::select(
        gene_id, gene_name, logFC, logCPM, PValue, FDR, everything()
      ) %&gt;%
      mutate(
        contrast = x,
        bonfP = p.adjust(PValue, &quot;bonf&quot;),
        DE = FDR &lt; alpha
        # DE = case_when(
        #   bonfP &lt; alpha ~ TRUE,
        #   FDR &lt; alpha &amp; abs(logFC) &gt; minLfc ~ TRUE
        # ),
        # DE = ifelse(is.na(DE), FALSE, DE)
      )
  }, simplify = FALSE)</code></pre>
<p>With this criteria, the following genes were classified as DE:</p>
<pre class="r"><code>for (ind in seq_along(topTables)) {
  print({
    topTables[[ind]] %&gt;%
      dplyr::filter(DE) %&gt;%
      dplyr::select(gene_id, gene_name, chromosome, logFC, logCPM, PValue, FDR, bonfP) %&gt;%
      mutate(
        logCPM = formatC(logCPM, digits = 2, format = &quot;f&quot;),
        logFC = formatC(logFC, digits = 2, format = &quot;f&quot;),
        PValue = formatC(PValue, digits = 2, format = &quot;e&quot;),
        FDR = formatC(FDR, digits = 2, format = &quot;e&quot;),
        bonfP = formatC(bonfP, digits = 2, format = &quot;e&quot;),
      ) %&gt;%
      kable(
        align = &quot;l&quot;,
        caption = paste0(names(topTables[ind]), &quot;. The &quot;, nrow(.), &quot; differentially expressed genes.&quot;)
      ) %&gt;%
      kable_styling(
        bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
      )
  })
}</code></pre>
<pre class="r"><code>deCols &lt;- c(`TRUE` = &quot;red&quot;, `FALSE` = &quot;grey50&quot;)</code></pre>
</div>
<div id="bias-checks" class="section level2">
<h2>Bias checks</h2>
<div id="ma-plots" class="section level3">
<h3>MA plots</h3>
<pre class="r"><code>topTables %&gt;%
  bind_rows() %&gt;%
  arrange(DE) %&gt;%
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  # geom_text_repel(
  #   aes(label = gene_name, colour = DE),
  #   data = . %&gt;% dplyr::filter(DE),
  #   max.overlaps = 20
  # ) +
  geom_smooth(se = FALSE) +
  facet_wrap(~contrast, ncol= 2) +
  scale_y_continuous(breaks = seq(-8, 8, by = 2)) +
  scale_colour_manual(values = deCols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-10-1.png" alt="*MA plots checking for logFC artefacts across the range of expression values. The average logFC appears consistant across all expression values for both comparisons*" width="768" />
<p class="caption">
<em>MA plots checking for logFC artefacts across the range of expression
values. The average logFC appears consistant across all expression
values for both comparisons</em>
</p>
</div>
</div>
<div id="gc-content" class="section level3">
<h3>GC content</h3>
<pre class="r"><code>topTables %&gt;%
  bind_rows() %&gt;%
  mutate(stat = -sign(logFC)*log10(PValue)) %&gt;%
  ggplot(aes(gc_content, stat)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  geom_smooth(se = FALSE) +
  facet_wrap(~contrast, ncol = 2)  +
  # geom_text_repel(
  #   aes(label = gene_name, colour = DE),
  #   data = . %&gt;% dplyr::filter(DE)
  # ) +
  labs(
    x = &quot; Gene GC content (%)&quot;,
    y = &quot;Ranking Statistic&quot;
  ) +
  coord_cartesian(ylim = c(-10, 10)) +
  scale_colour_manual(values = deCols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-11-1.png" alt="*GC content bias for differential expression. GC content is shown against the ranking statistic, calculated by -log10(p) multiple by the sign of logFC. Differentially expressed genes are labelled red. Small biases were observed for both comparisons. In particular, the fAI genotype shows a well-documented pattern of GC bias, whereby both low and high GC proportions are affected.*" width="768" />
<p class="caption">
<em>GC content bias for differential expression. GC content is shown
against the ranking statistic, calculated by -log10(p) multiple by the
sign of logFC. Differentially expressed genes are labelled red. Small
biases were observed for both comparisons. In particular, the fAI
genotype shows a well-documented pattern of GC bias, whereby both low
and high GC proportions are affected.</em>
</p>
</div>
</div>
<div id="gene-length" class="section level3">
<h3>Gene length</h3>
<pre class="r"><code>topTables %&gt;%
  bind_rows() %&gt;%
  mutate(stat = -sign(logFC)*log10(PValue)) %&gt;%
  ggplot(aes(length, stat)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  geom_smooth(se = FALSE) +
  facet_wrap(~contrast, ncol = 2)  +
  # geom_text_repel(
  #   aes(label = gene_name, colour = DE),
  #   data = . %&gt;% dplyr::filter(DE)
  # ) +
  labs(
    x = &quot;GC length&quot;,
    y = &quot;Ranking Statistic&quot;
  ) +
  coord_cartesian(ylim = c(-10, 10)) +
  scale_x_log10(labels = comma) +
  scale_colour_manual(values = deCols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-12-1.png" alt="*Gene length bias for differential expression. Gene length is shown against the ranking statistic, calculated by -log10(p) multiple by the sign of logFC. Differentially expressed genes are labelled red. Minimal bias is observed for the fAD genotype, but the fAI genotype shows a bias for larger genes.*" width="768" />
<p class="caption">
<em>Gene length bias for differential expression. Gene length is shown
against the ranking statistic, calculated by -log10(p) multiple by the
sign of logFC. Differentially expressed genes are labelled red. Minimal
bias is observed for the fAD genotype, but the fAI genotype shows a bias
for larger genes.</em>
</p>
</div>
</div>
</div>
</div>
<div id="conditional-quantile-normalisation" class="section level1">
<h1>Conditional Quantile Normalisation</h1>
<p>Conditional Quantile Normalisation (CQN) is a procedure that can
account for GC content and length biases in differential expression and
is implemented with package <code>cqn</code>. It does this by making by
calculating offset values that can be incorporated into the GLM. Both GC
content and length were noted as being of concern and were therefore
used to calculate offsets. The resulting <code>glm.offset</code> values
from the output of <code>cqn</code> were added to the original
<code>DGEList</code> object.</p>
<pre class="r"><code># cqn &lt;- cqn(
#   counts = dgeList$counts,
#   x = dgeList$genes$gc_content,
#   lengths = rep(1000, times = nrow(dgeList)),
#   sizeFactors = dgeList$samples$lib.size,
#   lengthMethod = &quot;fixed&quot;
# )
cqn &lt;- cqn(
  counts = dgeList$counts,
  x = dgeList$genes$gc_content,
  lengths = dgeList$genes$length,
  sizeFactors = dgeList$samples$lib.size
)</code></pre>
<pre class="r"><code>par(mfrow = c(1, 2))
cqnplot(cqn, n = 1, xlab = &quot;GC Content&quot;, col = genoCols)
cqnplot(cqn, n = 2, xlab = &quot;Length&quot;, col = genoCols)
legend(
  &quot;bottomright&quot;,
  legend = levels(metadata$genotype),
  col = genoCols,
  lty = 1
)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-14-1.png" alt="*Model fits for GC content and gene length under the CQN model. *" width="768" />
<p class="caption">
<em>Model fits for GC content and gene length under the CQN model. </em>
</p>
</div>
<pre class="r"><code>par(mfrow = c(1, 1))</code></pre>
<pre class="r"><code>dgeList$offset &lt;- cqn$glm.offset</code></pre>
<div id="pca-1" class="section level2">
<h2>PCA</h2>
<pre class="r"><code>cpm_cqn &lt;- cqn %&gt;%
  with(y + offset)</code></pre>
<pre class="r"><code>pca_cqn &lt;- cpm_cqn %&gt;%
  t() %&gt;%
  prcomp() 
pcaVars &lt;- percent_format(0.1)(summary(pca_cqn)$importance[&quot;Proportion of Variance&quot;,])</code></pre>
<pre class="r"><code>pca_cqn$x %&gt;%
  as.data.frame() %&gt;%
  rownames_to_column(&quot;sample&quot;) %&gt;%
  left_join(metadata) %&gt;%
  as_tibble() %&gt;%
  ggplot(aes(PC1, PC2, colour = genotype, fill = genotype, label = sample)) +
  geom_point(size = 2) +
  geom_text_repel(show.legend = FALSE) +
  stat_ellipse(geom = &quot;polygon&quot;, alpha = 0.05, show.legend = FALSE) +
  guides(fill = &quot;none&quot;) +
  scale_colour_manual(values = genoCols) +
  labs(
    x = paste0(&quot;PC1 (&quot;, pcaVars[[&quot;PC1&quot;]], &quot;)&quot;),
    y = paste0(&quot;PC2 (&quot;, pcaVars[[&quot;PC2&quot;]], &quot;)&quot;),
    colour = &quot;Genotype&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-18-1.png" alt="*Principal Component Analysis of gene-level counts post CQN.*" width="768" />
<p class="caption">
<em>Principal Component Analysis of gene-level counts post CQN.</em>
</p>
</div>
<p>There was very little improvement in separation of genotype groups
after CQN, and PC1 still only accounted for 18.2% of the variation. The
x-axis of the PCA plot appears to be flipped, but this normal as the
signs of PC values are somewhat arbitrary.</p>
</div>
<div id="model-fitting-1" class="section level2">
<h2>Model fitting</h2>
<p>After incorporation of the offset values into the
<code>DGEList</code> object, dispersion estimates were calculated and
models were fitted as done pre CQN.</p>
<pre class="r"><code>alpha &lt;- 0.05
minLfc &lt;- 1
fit_cqn &lt;- estimateDisp(dgeList, design) %&gt;%
  glmFit()
topTables_cqn &lt;- colnames(contrasts) %&gt;%
  sapply(function(x){
    glmLRT(fit_cqn, contrast = contrasts[,x]) %&gt;%
      topTags(n = Inf) %&gt;%
      .[[&quot;table&quot;]] %&gt;%
      as_tibble() %&gt;%
      arrange(PValue) %&gt;%
      dplyr::select(
        gene_id, gene_name, logFC, logCPM, PValue, FDR, everything()  
      ) %&gt;%
      mutate(
        contrast = x,
        bonfP = p.adjust(PValue, &quot;bonf&quot;),
        DE = FDR &lt; alpha
        # DE = case_when(
        #   bonfP &lt; alpha ~ TRUE,
        #   FDR &lt; alpha &amp; abs(logFC) &gt; minLfc ~ TRUE
        # ),
        # DE = ifelse(is.na(DE), FALSE, DE)
      )
  }, simplify = FALSE)</code></pre>
<pre class="r"><code>lapply(seq_along(topTables_cqn), function(ind){
  topTables_cqn[[ind]] %&gt;%
    dplyr::filter(DE) %&gt;% 
    dplyr::select(gene_id, gene_name, chromosome, logFC, logCPM, PValue, FDR, bonfP) %&gt;%
    mutate(
      logCPM = formatC(logCPM, digits = 2, format = &quot;f&quot;),
      logFC = formatC(logFC, digits = 2, format = &quot;f&quot;),
      PValue = formatC(PValue, digits = 2, format = &quot;e&quot;),
      FDR = formatC(FDR, digits = 2, format = &quot;e&quot;),
      bonfP = formatC(bonfP, digits = 2, format = &quot;e&quot;),
    ) %&gt;%
    kable(
      align = &quot;l&quot;,
      caption = paste0(names(topTables[ind]), &quot;. The &quot;, nrow(.), &quot; differentially expressed genes.&quot;)
    ) %&gt;%
    kable_styling(
      bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;, &quot;condensed&quot;, &quot;responsive&quot;)
    )
})</code></pre>
<pre><code>## [[1]]
## 
## [[2]]
## 
## [[3]]
## 
## [[4]]</code></pre>
</div>
<div id="bias-checks-1" class="section level2">
<h2>Bias checks</h2>
<div id="ma-plots-1" class="section level3">
<h3>MA plots</h3>
<pre class="r"><code>topTables_cqn %&gt;%
  bind_rows() %&gt;%
  arrange(DE) %&gt;%
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  # geom_text_repel(
  #   aes(label = gene_name, colour = DE),
  #   data = . %&gt;% dplyr::filter(DE)
  # ) +
  geom_smooth(se = FALSE) +
  facet_wrap(~contrast, ncol = 2) +
  scale_y_continuous(breaks = seq(-8, 8, by = 2)) +
  scale_colour_manual(values = deCols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-21-1.png" alt="*MA plots checking for logFC artefacts across the range of expression values. The average logFC remains consistant across all expression values for both comparisons after CQN*" width="768" />
<p class="caption">
<em>MA plots checking for logFC artefacts across the range of expression
values. The average logFC remains consistant across all expression
values for both comparisons after CQN</em>
</p>
</div>
</div>
<div id="gc-content-1" class="section level3">
<h3>GC content</h3>
<pre class="r"><code>topTables_cqn %&gt;%
  bind_rows() %&gt;%
  mutate(stat = -sign(logFC)*log10(PValue)) %&gt;%
  ggplot(aes(gc_content, stat)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  geom_smooth(se = FALSE) +
  facet_wrap(~contrast, ncol = 2)  +
  # geom_text_repel(
  #   aes(label = gene_name, colour = DE),
  #   data = . %&gt;% dplyr::filter(DE)
  # ) +
  labs(
    x = &quot;Gene GC content (%)&quot;,
    y = &quot;Ranking Statistic&quot;
  ) +
  coord_cartesian(ylim = c(-10, 10)) +
  scale_colour_manual(values = deCols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-22-1.png" alt="*GC content bias for differential expression. GC content is shown against the ranking statistic, calculated by -log10(p) multiple by the sign of logFC. Differentially expressed genes are labelled red. No improvement of GC bias was observed post-CQN. In fact, GC bias appeared worse than pre-CQN.*" width="768" />
<p class="caption">
<em>GC content bias for differential expression. GC content is shown
against the ranking statistic, calculated by -log10(p) multiple by the
sign of logFC. Differentially expressed genes are labelled red. No
improvement of GC bias was observed post-CQN. In fact, GC bias appeared
worse than pre-CQN.</em>
</p>
</div>
</div>
<div id="gene-length-1" class="section level3">
<h3>Gene length</h3>
<pre class="r"><code>topTables_cqn %&gt;%
  bind_rows() %&gt;%
  mutate(stat = -sign(logFC)*log10(PValue)) %&gt;%
  ggplot(aes(length, stat)) +
  geom_point(aes(colour = DE), alpha = 0.4) +
  geom_smooth(se = FALSE) +
  facet_wrap(~contrast, ncol = 2)  +
  # geom_text_repel(
  #   aes(label = gene_name, colour = DE),
  #   data = . %&gt;% dplyr::filter(DE)
  # ) +
  labs(
    x = &quot;GC length&quot;,
    y = &quot;Ranking Statistic&quot;
  ) +
  coord_cartesian(ylim = c(-10, 10)) +
  scale_x_log10(labels = comma) +
  scale_colour_manual(values = deCols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-23-1.png" alt="*Gene length bias for differential expression. Gene length is shown against the ranking statistic, calculated by -log10(p) multiple by the sign of logFC. Differentially expressed genes are labelled red. No improvement of gene length bias was observed for either genotype post-CQN.*" width="768" />
<p class="caption">
<em>Gene length bias for differential expression. Gene length is shown
against the ranking statistic, calculated by -log10(p) multiple by the
sign of logFC. Differentially expressed genes are labelled red. No
improvement of gene length bias was observed for either genotype
post-CQN.</em>
</p>
</div>
</div>
</div>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<pre class="r"><code>topTables_chosen &lt;- topTables_cqn</code></pre>
<pre class="r"><code>topTables_chosen %&gt;%
  bind_rows() %&gt;%
  ggplot(aes(logFC, -log10(PValue), colour = DE)) +
  geom_point(alpha = 0.4) +
  # geom_text_repel(
  #   aes(label = gene_name),
  #   size = 3,
  #   data = . %&gt;% dplyr::filter(DE)
  # ) +
  # geom_text_repel(
  #   aes(label = gene_name),
  #   size = 3,
  #   data = . %&gt;% dplyr::filter(!DE, logFC &lt; -3)
  # ) +
  facet_wrap(~contrast, ncol = 2) +
  scale_colour_manual(values = deCols) +
  scale_x_continuous(breaks = seq(-8, 8, by = 2)) +
  coord_cartesian(ylim = c(0, 10)) +
  labs(y = log10plab) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="de_files/figure-html/unnamed-chunk-25-1.png" alt="*Volcano plots showing -log10(p) against logFC. Genes determined to be DE are coloured red.*" width="768" />
<p class="caption">
<em>Volcano plots showing -log10(p) against logFC. Genes determined to
be DE are coloured red.</em>
</p>
</div>
<pre class="r"><code>tt &lt;- topTables_chosen$APOE2v3_female
man &lt;- tt %&gt;%
  mutate(mid = (start + end) / 2) %&gt;%
  dplyr::select(gene_id, gene_name, chromosome, mid, FDR) %&gt;%
  dplyr::filter(!(chromosome %in% c(&quot;X&quot;, &quot;Y&quot;, &quot;MT&quot;))) %&gt;%
  group_by(chromosome) %&gt;%
  summarise(chrLen = max(mid)) %&gt;%
  mutate(chrSt = cumsum(chrLen)-chrLen) %&gt;%
  dplyr::select(-chrLen) %&gt;% 
  left_join(tt) %&gt;%
  mutate(mid = (start + end) / 2) %&gt;%
  mutate(midCum = chrSt + mid) %&gt;%
  mutate(
    colour = ifelse(chromosome %in% primary_chrs[c(TRUE, FALSE)], &quot;a&quot;, &quot;b&quot;),
    colour = ifelse(chromosome == &quot;7&quot;, &quot;c&quot;, colour)
  )
axis &lt;- man %&gt;%
  group_by(chromosome) %&gt;%
  summarise(center = (max(midCum) + min(midCum)) / 2)
ggplot(man, aes(x = midCum, y = -log10(PValue))) +
  geom_point(
    data = man %&gt;%
      dplyr::filter(!DE),
    aes(colour = colour),
    shape = 20,
    alpha = 1,
    size = 2
  ) +
  scale_colour_manual(values = c(&quot;grey50&quot;, &quot;grey80&quot;, &quot;red&quot;)) +
  geom_point(
    data = man %&gt;%
      dplyr::filter(DE),
    aes(fill = colour),
    colour = &quot;black&quot;,
    shape = 21,
    alpha = 1,
    size = 1.5
  ) +
  scale_fill_manual(values = c(&quot;grey50&quot;, &quot;grey80&quot;, &quot;red&quot;)) +
  # scale_shape_manual(values = c(20, 21)) +
  scale_x_continuous(label = axis$chromosome, breaks = axis$center) +
  scale_y_continuous(expand = c(0, 0)) +
  coord_cartesian(clip = &quot;off&quot;) +
  labs(
    # title = paste0(unique(topTable$contrast)),
    x = &quot;Chromosome&quot;,
    y = expression(paste(-log[10], &quot;(p)&quot;))
  ) +
  # geom_text_repel(
  #   data = dplyr::filter(man, DE),
  #   aes(label = gene_name, colour = colour),
  #   size = 2.5,
  #   max.overlaps = 50
  # ) +
  theme(
    legend.position=&quot;none&quot;,
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank()
  )</code></pre>
<p><img src="de_files/figure-html/unnamed-chunk-26-1.png" width="768" style="display: block; margin: auto;" /></p>
<div id="data-export" class="section level2">
<h2>Data export</h2>
<p>Final results were exported as Rds files for further analysis.</p>
<pre class="r"><code>saveRDS(topTables, here(&quot;files/topTables.Rds&quot;))
saveRDS(topTables_cqn, here(&quot;files/topTables_cqn.Rds&quot;))</code></pre>
<pre class="r"><code>sessionInfo() %&gt;%
  pander()</code></pre>
<p><strong>R version 4.2.2 Patched (2022-11-10 r83330)</strong></p>
<p><strong>Platform:</strong> x86_64-pc-linux-gnu (64-bit)</p>
<p><strong>locale:</strong> <em>LC_CTYPE=en_AU.UTF-8</em>,
<em>LC_NUMERIC=C</em>, <em>LC_TIME=en_AU.UTF-8</em>,
<em>LC_COLLATE=en_AU.UTF-8</em>, <em>LC_MONETARY=en_AU.UTF-8</em>,
<em>LC_MESSAGES=en_AU.UTF-8</em>, <em>LC_PAPER=en_AU.UTF-8</em>,
<em>LC_NAME=C</em>, <em>LC_ADDRESS=C</em>, <em>LC_TELEPHONE=C</em>,
<em>LC_MEASUREMENT=en_AU.UTF-8</em> and <em>LC_IDENTIFICATION=C</em></p>
<p><strong>attached base packages:</strong> <em>splines</em>,
<em>grid</em>, <em>stats4</em>, <em>stats</em>, <em>graphics</em>,
<em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em>
and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>gtools(v.3.9.4)</em>,
<em>htmltools(v.0.5.4)</em>, <em>DT(v.0.27)</em>,
<em>cqn(v.1.44.0)</em>, <em>quantreg(v.5.94)</em>,
<em>SparseM(v.1.81)</em>, <em>preprocessCore(v.1.60.2)</em>,
<em>nor1mix(v.1.3-0)</em>, <em>mclust(v.6.0.0)</em>,
<em>pheatmap(v.1.0.12)</em>, <em>fgsea(v.1.24.0)</em>,
<em>edgeR(v.3.40.2)</em>, <em>tictoc(v.1.1)</em>,
<em>future.apply(v.1.10.0)</em>, <em>future(v.1.31.0)</em>,
<em>ensembldb(v.2.22.0)</em>, <em>AnnotationFilter(v.1.22.0)</em>,
<em>GenomicFeatures(v.1.50.4)</em>, <em>AnnotationDbi(v.1.60.0)</em>,
<em>limma(v.3.54.1)</em>, <em>readxl(v.1.4.2)</em>,
<em>msigdbr(v.7.5.1)</em>, <em>zoo(v.1.8-11)</em>,
<em>Gviz(v.1.42.1)</em>, <em>VariantAnnotation(v.1.44.1)</em>,
<em>Rsamtools(v.2.14.0)</em>, <em>Biostrings(v.2.66.0)</em>,
<em>XVector(v.0.38.0)</em>, <em>SummarizedExperiment(v.1.28.0)</em>,
<em>Biobase(v.2.58.0)</em>, <em>GenomicRanges(v.1.50.2)</em>,
<em>GenomeInfoDb(v.1.34.9)</em>, <em>IRanges(v.2.32.0)</em>,
<em>S4Vectors(v.0.36.1)</em>, <em>MatrixGenerics(v.1.10.0)</em>,
<em>matrixStats(v.0.63.0)</em>, <em>viridis(v.0.6.2)</em>,
<em>viridisLite(v.0.4.1)</em>, <em>rmarkdown(v.2.20)</em>,
<em>pander(v.0.6.5)</em>, <em>ggpubr(v.0.6.0)</em>,
<em>RColorBrewer(v.1.1-3)</em>, <em>ggrepel(v.0.9.3)</em>,
<em>kableExtra(v.1.3.4)</em>, <em>scales(v.1.2.1)</em>,
<em>AnnotationHub(v.3.6.0)</em>, <em>BiocFileCache(v.2.6.1)</em>,
<em>dbplyr(v.2.3.0)</em>, <em>BiocGenerics(v.0.44.0)</em>,
<em>magrittr(v.2.0.3)</em>, <em>lubridate(v.1.9.2)</em>,
<em>forcats(v.1.0.0)</em>, <em>stringr(v.1.5.0)</em>,
<em>dplyr(v.1.1.0)</em>, <em>purrr(v.1.0.1)</em>,
<em>readr(v.2.1.4)</em>, <em>tidyr(v.1.3.0)</em>,
<em>tibble(v.3.1.8)</em>, <em>ggplot2(v.3.4.1)</em>,
<em>tidyverse(v.2.0.0)</em> and <em>here(v.1.0.1)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong>
<em>utf8(v.1.2.3)</em>, <em>tidyselect(v.1.2.0)</em>,
<em>RSQLite(v.2.3.0)</em>, <em>htmlwidgets(v.1.6.1)</em>,
<em>BiocParallel(v.1.32.5)</em>, <em>munsell(v.0.5.0)</em>,
<em>codetools(v.0.2-19)</em>, <em>interp(v.1.1-3)</em>,
<em>withr(v.2.5.0)</em>, <em>colorspace(v.2.1-0)</em>,
<em>filelock(v.1.0.2)</em>, <em>highr(v.0.10)</em>,
<em>knitr(v.1.42)</em>, <em>rstudioapi(v.0.14)</em>,
<em>ggsignif(v.0.6.4)</em>, <em>listenv(v.0.9.0)</em>,
<em>labeling(v.0.4.2)</em>, <em>GenomeInfoDbData(v.1.2.9)</em>,
<em>farver(v.2.1.1)</em>, <em>bit64(v.4.0.5)</em>,
<em>rprojroot(v.2.0.3)</em>, <em>parallelly(v.1.34.0)</em>,
<em>vctrs(v.0.5.2)</em>, <em>generics(v.0.1.3)</em>,
<em>xfun(v.0.37)</em>, <em>biovizBase(v.1.46.0)</em>,
<em>timechange(v.0.2.0)</em>, <em>R6(v.2.5.1)</em>,
<em>locfit(v.1.5-9.7)</em>, <em>bitops(v.1.0-7)</em>,
<em>cachem(v.1.0.6)</em>, <em>DelayedArray(v.0.24.0)</em>,
<em>assertthat(v.0.2.1)</em>, <em>vroom(v.1.6.1)</em>,
<em>promises(v.1.2.0.1)</em>, <em>BiocIO(v.1.8.0)</em>,
<em>nnet(v.7.3-18)</em>, <em>gtable(v.0.3.1)</em>,
<em>globals(v.0.16.2)</em>, <em>MatrixModels(v.0.5-1)</em>,
<em>rlang(v.1.0.6)</em>, <em>systemfonts(v.1.0.4)</em>,
<em>rtracklayer(v.1.58.0)</em>, <em>rstatix(v.0.7.2)</em>,
<em>lazyeval(v.0.2.2)</em>, <em>dichromat(v.2.0-0.1)</em>,
<em>broom(v.1.0.3)</em>, <em>plyranges(v.1.18.0)</em>,
<em>checkmate(v.2.1.0)</em>, <em>BiocManager(v.1.30.20)</em>,
<em>yaml(v.2.3.7)</em>, <em>abind(v.1.4-5)</em>,
<em>backports(v.1.4.1)</em>, <em>httpuv(v.1.6.9)</em>,
<em>Hmisc(v.4.8-0)</em>, <em>tools(v.4.2.2)</em>,
<em>ellipsis(v.0.3.2)</em>, <em>jquerylib(v.0.1.4)</em>,
<em>Rcpp(v.1.0.10)</em>, <em>base64enc(v.0.1-3)</em>,
<em>progress(v.1.2.2)</em>, <em>zlibbioc(v.1.44.0)</em>,
<em>RCurl(v.1.98-1.10)</em>, <em>prettyunits(v.1.1.1)</em>,
<em>rpart(v.4.1.19)</em>, <em>deldir(v.1.0-6)</em>,
<em>cowplot(v.1.1.1)</em>, <em>cluster(v.2.1.4)</em>,
<em>data.table(v.1.14.8)</em>, <em>ProtGenerics(v.1.30.0)</em>,
<em>hms(v.1.1.2)</em>, <em>mime(v.0.12)</em>, <em>evaluate(v.0.20)</em>,
<em>xtable(v.1.8-4)</em>, <em>XML(v.3.99-0.13)</em>,
<em>jpeg(v.0.1-10)</em>, <em>gridExtra(v.2.3)</em>,
<em>compiler(v.4.2.2)</em>, <em>biomaRt(v.2.54.0)</em>,
<em>crayon(v.1.5.2)</em>, <em>mgcv(v.1.8-41)</em>,
<em>later(v.1.3.0)</em>, <em>tzdb(v.0.3.0)</em>,
<em>Formula(v.1.2-4)</em>, <em>DBI(v.1.1.3)</em>,
<em>MASS(v.7.3-58.2)</em>, <em>rappdirs(v.0.3.3)</em>,
<em>babelgene(v.22.9)</em>, <em>Matrix(v.1.5-3)</em>,
<em>car(v.3.1-1)</em>, <em>cli(v.3.6.0)</em>,
<em>parallel(v.4.2.2)</em>, <em>pkgconfig(v.2.0.3)</em>,
<em>GenomicAlignments(v.1.34.0)</em>, <em>foreign(v.0.8-84)</em>,
<em>xml2(v.1.3.3)</em>, <em>svglite(v.2.1.1)</em>,
<em>bslib(v.0.4.2)</em>, <em>webshot(v.0.5.4)</em>,
<em>rvest(v.1.0.3)</em>, <em>digest(v.0.6.31)</em>,
<em>fastmatch(v.1.1-3)</em>, <em>cellranger(v.1.1.0)</em>,
<em>htmlTable(v.2.4.1)</em>, <em>restfulr(v.0.0.15)</em>,
<em>curl(v.5.0.0)</em>, <em>shiny(v.1.7.4)</em>,
<em>rjson(v.0.2.21)</em>, <em>nlme(v.3.1-162)</em>,
<em>lifecycle(v.1.0.3)</em>, <em>jsonlite(v.1.8.4)</em>,
<em>carData(v.3.0-5)</em>, <em>BSgenome(v.1.66.3)</em>,
<em>fansi(v.1.0.4)</em>, <em>pillar(v.1.8.1)</em>,
<em>lattice(v.0.20-45)</em>, <em>KEGGREST(v.1.38.0)</em>,
<em>fastmap(v.1.1.0)</em>, <em>httr(v.1.4.4)</em>,
<em>survival(v.3.5-3)</em>, <em>interactiveDisplayBase(v.1.36.0)</em>,
<em>glue(v.1.6.2)</em>, <em>png(v.0.1-8)</em>,
<em>BiocVersion(v.3.16.0)</em>, <em>bit(v.4.0.5)</em>,
<em>stringi(v.1.7.12)</em>, <em>sass(v.0.4.5)</em>,
<em>blob(v.1.2.3)</em>, <em>latticeExtra(v.0.6-30)</em> and
<em>memoise(v.2.0.1)</em></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
