<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lachlan Baer" />


<title>Variants Analysis</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">211001_APOE_Mm</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="de.html">DE</a>
</li>
<li>
  <a href="variants.html">Variants</a>
</li>
<li>
  <a href="enrichment.html">Enrichment</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/baerlachlan/211001_APOE_Mm">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Variants Analysis</h1>
<h3 class="subtitle">211001_APOE_Mm dataset</h3>
<h4 class="author">Lachlan Baer</h4>
<h4 class="date">06 October, 2022</h4>

</div>


<div id="setup" class="section level1">
<h1>Setup</h1>
<pre class="r"><code>suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(VariantAnnotation)
  library(Gviz)
  library(zoo)
  library(msigdbr)
  library(viridis)
})</code></pre>
<pre class="r"><code>if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores &lt;- availableCores() - 1</code></pre>
<pre class="r"><code>source(&quot;~/bioinformatics/bioToolkit/lbFuncs.R&quot;)</code></pre>
<pre class="r"><code>## Choose either local or remote project directory
# projDir &lt;- here()
projDir &lt;- &quot;/hpcfs/users/a1647910/211001_APOE_Mm&quot;</code></pre>
<pre class="r"><code>options(ucscChromosomeNames = FALSE)</code></pre>
<div id="ensdb" class="section level2">
<h2>EnsDb</h2>
<pre class="r"><code>ens_species &lt;- &quot;Mus musculus&quot;
ens_release &lt;- &quot;104&quot;
ens_assembly &lt;- &quot;GRCm39&quot;</code></pre>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(species == ens_species) %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;)
ahId &lt;- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb &lt;- ah[[ahId]]</code></pre>
<pre class="r"><code>chrInfo &lt;- getChromInfoFromEnsembl(ens_assembly, release = ens_release) %&gt;%
  dplyr::filter(coord_system == &quot;chromosome&quot;)
primary_chrs &lt;- chrInfo$name</code></pre>
<pre class="r"><code>genes &lt;- genes(ensDb, filter = SeqNameFilter(primary_chrs))
mcols(genes) &lt;- mcols(genes)[
  c(&quot;gene_id&quot;, &quot;gene_name&quot;, &quot;gene_biotype&quot;, &quot;entrezid&quot;)
]</code></pre>
<pre class="r"><code>exons &lt;- exonsBy(ensDb, by = &quot;gene&quot;, filter = SeqNameFilter(primary_chrs))</code></pre>
</div>
<div id="metadata" class="section level2">
<h2>Metadata</h2>
<pre class="r"><code>metadata &lt;- read_tsv(here(&quot;misc/SYNAPSE_METADATA_MANIFEST.tsv&quot;)) %&gt;%
  left_join(read_csv(here(&quot;misc/metaAPOE.csv&quot;))) %&gt;%
  dplyr::select(
    sample = specimenID, species, genotypeBackground, litter, dateBirth,
    dateDeath, genotype = Genotype, sex = Sex, age = Age, lane, basename = name,
    modelSystemName, individualID, study
  ) %&gt;%
  dplyr::filter(str_detect(sample, &quot;_3M_&quot;)) %&gt;%
  mutate(basename = str_remove(basename, &quot;.bam_R(1|2).fastq.gz&quot;)) %&gt;%
  distinct(sample, .keep_all = TRUE) %&gt;%
  mutate(
    group = as.factor(paste0(genotype, &quot;_&quot;, age, &quot;_&quot;, sex)),
    genotype = as.factor(genotype)
  ) %&gt;%
  dplyr::arrange(genotype, group)</code></pre>
<pre class="r"><code>genoCols &lt;- metadata$genotype %&gt;%
  unique() %&gt;%
  length() %&gt;%
  brewer.pal(&quot;Set1&quot;) %&gt;%
  setNames(unique(metadata$genotype))</code></pre>
<pre class="r"><code>samples_byGroup &lt;- metadata %&gt;%
  split(f = .$group) %&gt;%
  sapply(function(x){
    pull(x, sample)
  }, simplify = FALSE)</code></pre>
</div>
</div>
<div id="de-data" class="section level1">
<h1>DE data</h1>
<pre class="r"><code>dgeList &lt;- readRDS(here(&quot;files/dgeList.Rds&quot;))</code></pre>
<pre class="r"><code>topDE &lt;- readRDS(here(&quot;files/topTables_cqn.Rds&quot;)) %&gt;%
  sapply(function(x){
    dplyr::filter(x, chromosome %in% primary_chrs) %&gt;%
      droplevels()
  }, simplify = FALSE)</code></pre>
<pre class="r"><code>deGenes &lt;- sapply(topDE, dplyr::filter, DE, simplify = FALSE)</code></pre>
</div>
<div id="variation-between-groups" class="section level1">
<h1>Variation between groups</h1>
<pre class="r"><code>vcf_file &lt;- file.path(
  &quot;/hpcfs/users/a1647910/211001_APOE_Mm&quot;,
  &quot;analysis-variants/results/07_variants/6_select/all_samples.vcf.gz&quot;
)
svp &lt;- ScanVcfParam(info = &quot;&quot;, geno = c(&quot;GT&quot;, &quot;GQ&quot;))
vcf &lt;- suppressWarnings({
  readVcf(vcf_file, param = svp)
})</code></pre>
<pre class="r"><code>gr &lt;- rowRanges(vcf)
mcols(gr) &lt;- NULL
gr &lt;- gr %&gt;%
  as.data.frame() %&gt;%
  droplevels() %&gt;%
  GRanges()</code></pre>
<pre class="r"><code>unphase_GTs &lt;- function(x){
  str_replace(x, &quot;\\|&quot;, &quot;\\/&quot;)
}
gt &lt;- geno(vcf)$GT %&gt;%
  as.data.frame() %&gt;%
  mutate(across(everything(), unphase_GTs)) %&gt;%
  set_colnames(str_remove(colnames(.), &quot;\\.FCHJKY5BBXX.+&quot;))</code></pre>
<div id="fold-change" class="section level2">
<h2>Fold change</h2>
<pre class="r"><code>lfc_winSize &lt;- 21
lfc &lt;- lapply(topDE, function(x){
  x %&gt;%
    dplyr::select(
      # Don&#39;t include strand because allelic differences between groups is consistent between strands
      gene_id, gene_name, logFC, seqnames = chromosome, start, end, width
    ) %&gt;%
    dplyr::filter(seqnames %in% primary_chrs) %&gt;%
    mutate(abs.logFC = abs(logFC)) %&gt;%
    split(f = .$seqnames) %&gt;%
    lapply(function(chr){
      mutate(
        chr,
        lfc_rollmean = rollmean(abs.logFC, k = lfc_winSize, fill = NA),
        win_start = lag(start, n = (lfc_winSize - 1) / 2),
        win_end = lead(end, n = (lfc_winSize - 1) / 2),
        win_size = abs(win_end - win_start)
      )
    }) %&gt;%
    bind_rows() %&gt;%
    GRanges()
})</code></pre>
<div id="elastic-window-sizes" class="section level3">
<h3>Elastic window sizes</h3>
<pre class="r"><code>lapply(seq_along(lfc), function(ind){
  lfc[[ind]] %&gt;%
    as_tibble() %&gt;%
    dplyr::filter(!is.na(win_size)) %&gt;%
    mutate(
      comparison = factor(names(lfc[ind]), levels = names(lfc[1:4]))
    )
}) %&gt;%
  bind_rows() %&gt;%
  ggplot(aes(win_size)) +
  geom_histogram(
    bins = 100,
    colour = &quot;black&quot;,
    fill = &quot;grey70&quot;
  ) +
  scale_x_log10() +
  facet_wrap(~comparison, ncol = 2) +
  labs(
    title = paste0(
      &quot;Distribution of elastic window sizes generated from &quot;,
      lfc_winSize,
      &quot; genes&quot;
    ),
    x = &quot;Window size (bp)&quot;,
    y = &quot;Count&quot;
  )</code></pre>
<p><img src="variants_files/figure-html/plot_lfc_winSize-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="allelic-diversity" class="section level2">
<h2>Allelic diversity</h2>
<pre class="r"><code>allele_file &lt;- file.path(here(), &quot;files&quot;, &quot;allele_stats.Rds&quot;)</code></pre>
<pre class="r"><code>## Takes 15-20 mins to run so .Rds saved for convenience
if (!file.exists(allele_file)) {
  gr_tbl &lt;- as_tibble(gr)
  allele_counts &lt;- lapply(samples_byGroup, function(samples){
    apply(gt[,samples], 1, function(genos){
      alleles &lt;- genos %&gt;%
        as.character() %&gt;%
        str_split(&quot;/&quot;) %&gt;%
        unlist()
      tibble(
        n_called = sum(alleles != &quot;.&quot;),
        n_nocall = sum(alleles == &quot;.&quot;),
        n_0 = sum(alleles == &quot;0&quot;),
        n_1 = sum(alleles == &quot;1&quot;),
        n_2 = sum(alleles == &quot;2&quot;),
        n_3 = sum(alleles == &quot;3&quot;)
      )
    }) %&gt;%
      bind_rows()
  })
  allele_stats &lt;- lapply(allele_counts, function(counts){
    cbind(gr_tbl, counts) %&gt;%
      mutate(
        prop_ref = ifelse(n_called &gt; n_nocall, n_0 / n_called, NA),
        prop_called = n_called / (n_called + n_nocall)
      ) %&gt;%
      nest(allele_counts = c(n_called, n_nocall, n_0, n_1, n_2, n_3)) %&gt;%
      GRanges()
  })
  saveRDS(object = allele_stats, file = allele_file)
} else {
  allele_stats &lt;- readRDS(allele_file)
}</code></pre>
<pre class="r"><code>allele_props &lt;- lapply(allele_stats, function(stats){
  stats$allele_counts %&gt;%
    bind_rows() %&gt;%
    mutate(
      across(
        .cols = c(&quot;n_0&quot;, &quot;n_1&quot;, &quot;n_2&quot;, &quot;n_3&quot;),
        .fns = ~ ifelse(n_called &gt; n_nocall, .x / n_called, NA)
      )
    ) %&gt;%
    dplyr::select(-n_called, -n_nocall)
})</code></pre>
<pre class="r"><code>dist_winSize &lt;- 11
allele_dist_path &lt;- here(&quot;files&quot;, &quot;allele_dist.Rds&quot;)
if (!file.exists(allele_dist_path)) {
allele_dist &lt;- lapply(
  list(
    allele_props[c(&quot;APOE2_3M_female&quot;, &quot;APOE3_3M_female&quot;)],
    allele_props[c(&quot;APOE2_3M_male&quot;, &quot;APOE3_3M_male&quot;)],
    allele_props[c(&quot;APOE4_3M_female&quot;, &quot;APOE3_3M_female&quot;)],
    allele_props[c(&quot;APOE4_3M_male&quot;, &quot;APOE3_3M_male&quot;)]
  ),
  function(props){
    dist &lt;- vapply(seq(length(gr)), function(ind){
      rbind(props[[1]][ind,], props[[2]][ind,]) %&gt;%
        dist() %&gt;%
        as.numeric()
    }, numeric(1))
    gr %&gt;%
      as_tibble() %&gt;%
      mutate(dist = dist) %&gt;%
      dplyr::filter(!is.na(dist)) %&gt;%
      split(f = .$seqnames) %&gt;%
      lapply(function(chr){
        dplyr::mutate(
          chr,
          ## Something strange is happening with rollmean()
          ## Consecutive 0s are being reported as having small negative rollmeans
          ## Seems like a bug, but rollapply with the partial argument works
          # dist_rollmean = rollmean(dist, k = dist_winSize, fill = NA),
          dist_rollmean = rollapply(dist, width = dist_winSize, mean, partial = TRUE),
          diversity = dist_rollmean / sqrt(2),  # sqrt(2) is the maximum dist
          win_start = lag(start, n = (dist_winSize - 1) / 2),
          win_end = lead(start, n = (dist_winSize - 1) / 2),
          win_size = win_end - win_start
        ) %&gt;%
          dplyr::select(-c(dist))
      }) %&gt;%
      bind_rows() %&gt;%
      GRanges()
  }
) %&gt;%
  set_names(c(
    &quot;APOE2v3_female&quot;,
    &quot;APOE2v3_male&quot;,
    &quot;APOE4v3_female&quot;,
    &quot;APOE4v3_male&quot;
  ))
  saveRDS(allele_dist, allele_dist_path)
} else {
  allele_dist &lt;- readRDS(allele_dist_path)
}</code></pre>
<div id="dar-distribution" class="section level3">
<h3>DAR distribution</h3>
<pre class="r"><code>lapply(seq_along(allele_dist), function(ind){
  allele_dist[[ind]] %&gt;%
    split(seqnames(.)) %&gt;%
    lapply(function(chr){
      tibble(
        chromosome = as.vector(seqnames(chr)),
        div = chr$diversity
      )
    }) %&gt;%
    purrr::reduce(rbind) %&gt;%
    mutate(
      chromosome = factor(
        chromosome,
        # Draw mutant chr on top
        levels = unique(c(
          chromosome[chromosome != &quot;15&quot;],
          chromosome[chromosome == &quot;15&quot;]
        ))
      ),
      mutant_chromosome = chromosome == &quot;15&quot;
    ) %&gt;%
    ggplot(aes(x = div, group = chromosome, colour = mutant_chromosome)) +
    stat_ecdf() +
    scale_colour_manual(values = c(&quot;grey50&quot;, &quot;red&quot;)) +
    labs(
      title = names(allele_dist[ind]),
      x = &quot;DAR&quot;,
      y = &quot;F(DAR)&quot;
    )
})</code></pre>
<pre><code>## [[1]]</code></pre>
<p><img src="variants_files/figure-html/unnamed-chunk-2-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre><code>## 
## [[2]]</code></pre>
<p><img src="variants_files/figure-html/unnamed-chunk-2-2.png" width="768" style="display: block; margin: auto;" /></p>
<pre><code>## 
## [[3]]</code></pre>
<p><img src="variants_files/figure-html/unnamed-chunk-2-3.png" width="768" style="display: block; margin: auto;" /></p>
<pre><code>## 
## [[4]]</code></pre>
<p><img src="variants_files/figure-html/unnamed-chunk-2-4.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="elastic-window-sizes-1" class="section level3">
<h3>Elastic window sizes</h3>
<pre class="r"><code>lapply(seq_along(allele_dist), function(ind){
  allele_dist[[ind]] %&gt;%
    as_tibble() %&gt;%
    dplyr::filter(!is.na(win_size)) %&gt;%
    mutate(
      comparison = factor(names(allele_dist[ind]), levels = names(allele_dist))
    )
}) %&gt;%
  bind_rows() %&gt;%
  ggplot(aes(win_size)) +
  geom_histogram(
    bins = 100,
    colour = &quot;black&quot;,
    fill = &quot;grey70&quot;
  ) +
  scale_x_log10() +
  facet_wrap(~comparison, ncol = 2) +
  labs(
    title = paste0(
      &quot;Distribution of elastic window sizes generated from &quot;,
      dist_winSize,
      &quot; SNVs&quot;
    ),
    x = &quot;Window size (bp)&quot;,
    y = &quot;Count&quot;
  )</code></pre>
<p><img src="variants_files/figure-html/plot_dist_winSize-1.png" width="768" style="display: block; margin: auto;" /></p>
<pre class="r"><code>dist_winRanges &lt;- lapply(allele_dist, function(x){
  x %&gt;%
    as_tibble() %&gt;%
    dplyr::filter(!is.na(win_size)) %&gt;%
    dplyr::select(
      seqnames, start = win_start, end = win_end, width = win_size, strand,
      dist_rollmean, diversity
    ) %&gt;%
    split(f = .$seqnames) %&gt;%
    ## Extend the start and ends so that the whole chromosome is covered
    lapply(function(chr){
      seqname &lt;- unique(chr$seqnames)
      chr$start[1] &lt;- 1
      chr$width[1] &lt;- chr$end[1] - chr$start[1]
      last_ind &lt;- nrow(chr)
      chr$end[last_ind] &lt;- chrInfo$length[chrInfo$name == seqname]
      chr$width[last_ind] &lt;- chr$end[last_ind] - chr$start[last_ind]
      chr
    }) %&gt;%
    bind_rows() %&gt;%
    GRanges()
})</code></pre>
<pre class="r"><code>dist_winRanges_path &lt;- file.path(here(), &quot;files&quot;, &quot;dist_winRanges.Rds&quot;)
if (!file.exists(dist_winRanges_path)) {
  saveRDS(dist_winRanges, dist_winRanges_path)
}</code></pre>
</div>
<div id="mastergr" class="section level3">
<h3>masterGR</h3>
<pre class="r"><code>## Takes ~5 mins to run so .Rds saved for convenience
## Note a few genes are lost that do not fall within a diversity range
masterGR_path &lt;- here(&quot;files/masterGR.Rds&quot;)
if (!file.exists(masterGR_path)) {
  masterGR &lt;- map2(dist_winRanges, topDE, function(winRanges, tt){
    overlaps &lt;- findOverlaps(GRanges(dgeList$genes), winRanges)
    lapply(unique(queryHits(overlaps)), function(query){
      subjects &lt;- subjectHits(overlaps)[queryHits(overlaps) == query]
      diversity &lt;- winRanges$diversity[subjects] %&gt;%
        mean()
      dgeList$genes[query,] %&gt;%
        mutate(diversity = diversity)
    }) %&gt;%
      bind_rows() %&gt;%
      left_join(tt[,c(&quot;gene_id&quot;, &quot;logFC&quot;, &quot;PValue&quot;, &quot;FDR&quot;, &quot;DE&quot;)]) %&gt;%
      GRanges()
  })
  saveRDS(masterGR, masterGR_path)
} else {
  masterGR &lt;- readRDS(masterGR_path)
}</code></pre>
</div>
<div id="visualisation" class="section level3">
<h3>Visualisation</h3>
<pre class="r"><code>plot_allele_dist &lt;- function(
  chr, genes = GRanges(), mutation = GRanges(),
  dist, dist_type = &quot;l&quot;, dist_trans = 1, highlight_genes = TRUE,
  lfc = NULL, lfc_type = &quot;heatmap&quot;, lfc_trans = 1/3,
  title
){
  tracks &lt;- list()

  if (length(mutation)) {
    mutation &lt;- mutation %&gt;%
      plyranges::filter(seqnames == chr) %&gt;%
      plyranges::mutate(symbol = gene_name)
    mutant_track &lt;- GeneRegionTrack(
      range = mutation,
      transcriptAnnotation = &quot;symbol&quot;,
      col = &quot;white&quot;,
      fill = &quot;white&quot;,
      showTranscriptId = TRUE,
      fontcolor.group = &quot;red&quot;,
      cex.group = 0.6,
      size = 0.3
    )
    tracks &lt;- append(tracks, mutant_track)
  }

  axis_track &lt;- GenomeAxisTrack(
    add53 = TRUE,
    add35 = TRUE,
    name = paste0(&quot;Chr&quot;, chr),
    showTitle = TRUE
  )
  if (length(mutation)) {
    axis_track &lt;- HighlightTrack(
      trackList = list(axis_track),
      range = mutation
    )
  }
  tracks &lt;- append(tracks, axis_track)

  if (length(genes)) {
    genes &lt;- genes %&gt;%
      plyranges::filter(seqnames == chr) %&gt;%
      plyranges::mutate(symbol = gene_name)
    gene_track &lt;- GeneRegionTrack(
      range = genes,
      name = &quot;DE&quot;,
      transcriptAnnotation = &quot;symbol&quot;,
      showTranscriptId = TRUE,
      fontcolor.group = 1,
      cex.group = 0.5,
      size = 0.3
    )
    tracks &lt;- append(tracks, gene_track)
  }

  dist_track &lt;- dist[,&quot;diversity&quot;] %&gt;%
    plyranges::filter(seqnames == chr) %&gt;%
    DataTrack(
      type = dist_type,
      name = &quot;Between-group genetic dissimilarity&quot;,
      window = -1,
      windowSize = 1,
      col = &quot;grey30&quot;,
      col.axis = &quot;black&quot;,
      yTicksAt = seq(0, 1, 0.1),
      ylim = c(0, 1),
      transformation = function(x){x^(dist_trans)}
    )
  if (highlight_genes) {
    if (all(length(genes) &amp; length(mutation))) {
      ranges &lt;- c(genes, mutation)
    } else if (length(genes)) {
      ranges &lt;- genes
    } else if (length(mutation)) {
      ranges &lt;- mutation
    }
    dist_track &lt;- HighlightTrack(
      trackList = list(dist_track),
      range = c(genes, mutation),
      col = c(rep(&quot;#ffd1d1&quot;, length(genes)), rep(&quot;red&quot;, length(mutation))),
      fill = c(rep(&quot;#ffd1d1&quot;, length(genes)), rep(&quot;red&quot;, length(mutation)))
    )
  }
  tracks &lt;- append(tracks, dist_track)

  if (length(lfc)) {
    lfc_track &lt;- lfc[,&quot;lfc_rollmean&quot;] %&gt;%
      plyranges::filter(seqnames == chr) %&gt;%
      DataTrack(
        type = lfc_type,
        name = paste0(&quot;abs.logFC^(1/3)&quot;),
        size = 1.5,
        window = -1,
        windowSize = 1,
        col.axis = &quot;black&quot;,
        showColorBar = TRUE,
        gradient = viridis(100, option = &quot;D&quot;),
        transformation = function(x){x^(lfc_trans)}
      )
    tracks &lt;- append(tracks, lfc_track)
  }

  plotTracks(
    trackList = tracks,
    main = title,
    cex.main = 1,
    col.title = &quot;black&quot;,
    background.title = &quot;white&quot;
  )
}</code></pre>
<pre class="r"><code>chr &lt;- &quot;7&quot;
ind &lt;- 1
plot_allele_dist(
  chr = chr,
  genes = genes[deGenes[[ind]]$gene_id],
  mutation = genes[genes$gene_name == &quot;Apoe&quot;],
  dist = allele_dist[[ind]], dist_type = &quot;l&quot;, dist_trans = 1,
  lfc = lfc[[ind]], lfc_type = &quot;heatmap&quot;, lfc_trans = 1/3,
  # title = paste0(
  #   names(allele_dist[ind]),
  #   &quot; - dissimilarity win of &quot;, dist_winSize, &quot; SNVs&quot;,
  #   &quot; - logFC win of &quot;, lfc_winSize, &quot; genes&quot;
  # )
  title = &quot;&quot;
)</code></pre>
<p><img src="variants_files/figure-html/distPlot1-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Try quasi likelihood for DE testing Look at regulatory TFT datasets
throw away gene sets with no DE genes check previours chromosomal
enrichment analysis</p>
<p>Circos plot</p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
