---
title: "Variants Analysis"
subtitle: "211001_APOE_Mm dataset"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

```{r}
lapply(seq_along(allele_dist), function(ind){
  allele_dist[[ind]] %>%
    as_tibble() %>%
    dplyr::filter(!is.na(win_size)) %>%
    mutate(
      comparison = factor(names(allele_dist[ind]), levels = names(allele_dist))
    )
}) %>%
  bind_rows() %>%
  ggplot(aes(win_size)) +
  geom_histogram(
    bins = 100,
    colour = "black",
    fill = "grey70"
  ) +
  scale_x_log10() +
  facet_wrap(~comparison, ncol = 2) +
  labs(
    x = "Window size (bp)",
    y = "Count"
  )
```

```{r}
ggsave(
  "~/phd/publications/fisher_pub/figures/supp_1.png",
  device = "png",
  width = 18,
  height = 12,
  units = "cm"
)
```

```{r}
allele_dist[[1]] %>%
  split(seqnames(.)) %>%
  lapply(function(chr){
    tibble(
      chromosome = as.vector(seqnames(chr)),
      div = chr$diversity
    )
  }) %>%
  purrr::reduce(rbind) %>%
  mutate(
    chromosome = factor(
      chromosome,
      # Draw mutant chr on top
      levels = unique(c(
        chromosome[chromosome != "7"],
        chromosome[chromosome == "7"]
      ))
    ),
    mutant_chromosome = chromosome == "7"
  ) %>%
  ggplot(aes(x = div, group = chromosome, colour = mutant_chromosome)) +
  stat_ecdf() +
  scale_colour_manual(values = c("grey50", "red")) +
  labs(
    title = names(allele_dist[1]),
    x = "DAR",
    y = "F(DAR)"
  )
```

```{r}
ggsave(
  "~/phd/publications/fisher_pub/figures/supp_2.png",
  device = "png",
  width = 18,
  height = 8,
  units = "cm"
)
```